src/sqlfluff/core/rules/analysis/select_crawler.py

```diff
@@ -148,16 +65,170 @@ def get_wildcard_info(self) -> List[WildcardInfo]:
+        """Look up a CTE by name, in the current or any parent scope."""
+        cte = self.ctes.get(name)
+        if cte:
+            if pop:
+                del self.ctes[name]
+            return cte
+        if self.parent:
+            return self.parent.lookup_cte(name, pop)
+        else:
+            return None
+
+    def crawl_sources(
+        self, segment: BaseSegment, recurse_into=True, pop=False
+    ) -> Generator[Union[str, "Query"], None, None]:
+        """Find SELECTs, table refs, or value table function calls in segment.
+
+        For each SELECT, yield a list of Query objects. As we find table
+        references or function call strings, yield those.
+        """
+        found_nested_select = False
+        for seg in segment.recursive_crawl(
+            "table_reference",
+            "set_expression",
+            "select_statement",
+            recurse_into=recurse_into,
+        ):
+            if seg is segment:
+                # If we are starting with a select_statement, recursive_crawl()
+                # returns the statement itself. Skip that.
+                continue
+
+            if seg.is_type("table_reference"):
+                if not seg.is_qualified():
+                    cte = self.lookup_cte(seg.raw, pop=pop)
+                    if cte:
+                        # It's a CTE.
+                        yield cte
+                # It's an external table.
+                yield seg.raw
+            else:
+                assert seg.is_type("set_expression", "select_statement")
+                found_nested_select = True
+                crawler = SelectCrawler(seg, self.dialect, parent=self)
+                # We know this will pass because we specified parent=self above.
+                assert crawler.query_tree
+                yield crawler.query_tree
+        if not found_nested_select:
+            # If we reach here, the SELECT may be querying from a value table
+            # function, e.g. UNNEST(). For our purposes, this is basically the
+            # same as an external table. Return the "table" part as a string.
+            table_expr = segment.get_child("table_expression")
+            if table_expr:
+                yield table_expr.raw
+
+
+class SelectCrawler:
+    """Class for dependency analysis among parts of a query."""
+
+    def __init__(
+        self, segment: BaseSegment, dialect: Dialect, parent: Optional[Query] = None
+    ):
+        self.dialect: Dialect = dialect
+        self.query_tree: Optional[Query] = None
+
+        queries: List[Query] = []
+        pop_queries_for = []
+
+        def append_query(query):
+            if queries:
+                query.parent = queries[-1]
+            queries.append(query)
+            pop_queries_for.append(path[-1])
+
+        cte_name = None
+        for event, path in SelectCrawler.visit_segments(segment):
+            in_with = queries and queries[-1].query_type == QueryType.WithCompound
```

Comment:
@barrywhart quick question for my personal curiosity :)

If we are at the top level of this recursion (i.e. the statement segment) then queries is an empty list. Therefore `in_with` evaluates to False, which is expected.

My question is: The `queries[-1]` on its own would be an index error, yet the whole expression still evaluates successfully to False. In Python does it not bother evaluating the second part of an `and` expression if the first part evaluates to false (and hence the expression can't possibly be True)?

Reply:
That's right. When a language behaves this way, it's referred to as "short-circuit" evaluation, and it's a really convenient feature, because otherwise, you'd often have to chain together `if` statements to get the same effect.

Short-circuit evaluation is pretty common in programming languages. I know that C and C++ have it. I think Java does. For other languages, I'm not sure.

Reply:
More detail here: https://www.geeksforgeeks.org/short-circuiting-techniques-python/

Reply:
@barrywhart amazing, thanks for the link! Makes total sense, I was aware of the generator uses but will def be taking advantage of the boolean version going forwards 😄